FRFuel.cs

```diff
@@ -88,350 +58,156 @@ public FRFuel()
-
-                bone = vehicle.Bones[boneIndex];
-
-                if (bone.IsValid)
+                if (Game.PlayerPed.IsInVehicle() && (Game.PlayerPed.CurrentVehicle.Model.IsCar || Game.PlayerPed.CurrentVehicle.Model.IsBike || Game.PlayerPed.CurrentVehicle.Model.IsQuadbike)
+                    && Game.PlayerPed.CurrentVehicle.GetPedOnSeat(VehicleSeat.Driver) == Game.PlayerPed && Game.PlayerPed.CurrentVehicle.IsAlive)
                 {
-                    break;
-                }
-            }
+                    if (!Game.PlayerPed.CurrentVehicle.HasDecor(fuelLevelPropertyName)) //init fuel
+                    {
+                        var _vehicleMaxFuel = GetVehicleFuelCapacity(Game.PlayerPed.CurrentVehicle);
+                        SetVehicleFuelLevel(Game.PlayerPed.CurrentVehicle, RandomizeFuelLevel(_vehicleMaxFuel));
+                    }
 
-            if (bone == null)
-            {
-                return vehicle.Position;
-            }
+                    var _vehicleFuel = GetVehicleFuelLevel(Game.PlayerPed.CurrentVehicle);
 
-            return bone.Position;
-        }
+                    if (Game.PlayerPed.CurrentVehicle.IsEngineRunning && _vehicleFuel > 0f)//Consume Fuel (Algorithm may need tweeked due to removal of tightly coupling fuel usage to frame rates)
+                    {
+                        float normalizedRPMValue = (float)Math.Pow(Game.PlayerPed.CurrentVehicle.CurrentRPM, 1.5);
+                        float consumedFuel = 0f;
 
-        /// <summary>
-        /// Checks if vehicle is within range of activation
-        /// for any pump at current gas station
-        /// </summary>
-        /// <param name="vehicle"></param>
-        /// <returns></returns>
-        public bool IsVehicleNearAnyPump(Vehicle vehicle)
-        {
-            var fuelTankPos = GetVehicleTankPos(vehicle);
+                        consumedFuel += normalizedRPMValue * Config.GetInstance().FuelRPMImpact;
+                        consumedFuel += Game.PlayerPed.CurrentVehicle.Acceleration * Config.GetInstance().FuelAccelerationImpact;
+                        consumedFuel += Game.PlayerPed.CurrentVehicle.MaxTraction * Config.GetInstance().FuelTractionImpact;
+                        consumedFuel += Game.PlayerPed.CurrentVehicle.Speed * Config.GetInstance().FuelSpeedImpact;
+                        consumedFuel *= Config.GetInstance().FuelConsumptionRate;
 
-            foreach (var pump in GasStations.pumps[currentGasStationIndex])
-            {
-                if (Vector3.DistanceSquared(pump, fuelTankPos) <= 20f)
-                {
-                    return true;
+                        _vehicleFuel = Math.Max(0f, _vehicleFuel - consumedFuel);
+                        SetVehicleFuelLevel(Game.PlayerPed.CurrentVehicle, _vehicleFuel);
+                    }
                 }
-            }
 
-            return false;
+                await BaseScript.Delay(500);//Check fuel usage every .5 seconds
```

Comment:
It is too slow, better keep doing it every tick.

Reply:
I'm not sure that I agree that .5 is too slow. Many of your arguments for the code have been due to performance and consuming fuel every tick is not performance oriented. Any length of time chosen for delaying the fuel tick can be adjusted for by increasing consumption rates